library-best-practices/JLBP-6.md

```diff
@@ -0,0 +1,165 @@
+  that particular method. The larger the portion of a library that breaks
+  between major versions, the higher the value is from renaming the Java
+  package.
+- Whether making breaking changes or making no breaking changes, don't publish
+  the same classes under multiple Maven IDs. Or put another way: Once you have
+  published under a particular Maven ID, you are stuck with it until you rename
+  your Java package.
+
+Extended discussion
+-------------------
+
+Consider the following renaming scenario:
+
+- The Maven coordinates start as `g1:a1:1.0.0`
+  - With no rename of the Maven ID, the new major version's Maven coordinates
+    are `g1:a1:2.0.0`. (Only one of the versions will be present on the
+    classpath even if both are in the dependency tree.)
+  - With a rename of the Maven ID, the new major version's Maven coordinates are
+    `g2:a2:2.0.0`. (Both versions can be present on the classpath.)
+- The Java package starts as `com.google.a1`
+  - With no rename, the new major version's Java package is still
+    `com.google.a1`.
+  - With a rename, the new major version's Java package is `com.google.a2`.
+
+Given this scenario, here are the possible combinations of renamings:
+
+- **Don't rename Java package**:
+  - **Don't rename Maven ID (case 1)**: This approach can result in diamond
+    dependency conflicts because different branches of a dependency tree can
+    depend on different major versions, and the build system (Maven or Gradle)
+    will only choose one of them to use. Users are forced to change their code
+    between versions, but only for library surface that was changed.
+  - **Rename Maven ID (case 2)**: Users can easily pull in both the old jar
+    (`g1:a1:1.0.0`) and the new jar (`g2:a2:2.0.0`) accidentally through
+    transitive dependencies, because Maven treats the two artifacts as distinct,
+    but since they have classes with the same Java packages, they are loaded in
+    unpredictable ways, leading to runtime exceptions (for example
+    `ClassNotFoundException`). It is difficult for users to ensure that their
+    build tree only includes one of the artifacts. **NEVER DO THIS**.
+- **Rename Java package**:
+  - **Don't rename Maven ID (case 3)**: The classes from `g1:a1:1.0.0` and
+    `g1:a1:2.0.0` could technically be used together, but since they share the
+    Maven group ID and artifact ID, only one jar can be pulled in. Users are
+    forced to change their code if they want to add any references to the new
+    version. It is strictly better to either also rename the Maven ID (case 4)
+    or to keep the Maven ID and also bundle the old package (case 5), so that
+    the major versions can be used side by side.
+  - **Rename Maven ID (case 4)**: The two major versions can be used side by
+    side, allowing users to incrementally transition from the old to the new
+    version, or even use them side by side indefinitely if necessary. This is
+    similar to the approach taken by Go, described in
```

Comment:
This document's audience is Java developers. Are we just muddying the waters by talking about Go?

Reply:
Yeah, probably. I will remove it.